allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,274 @@
+        # (batch_size, enc_embedding_dim)
+        embedded_input = self._target_embedder(input_indices)
+        if self._use_attention:
+            if last_decoder_output is None or encoder_outputs is None:
+                raise ConfigurationError("Last decoder output and all encoder outputs are needed to compute "
+                                         "attention weights.")
+            # encoder_outputs : (batch_size, input_sequence_length, enc_output_dim)
+            # (batch_size, input_sequence_length)
+            input_weights = self._decoder_attention(last_decoder_output, encoder_outputs)
+            # (batch_size, enc_output_dim)
+            attended_input = weighted_sum(encoder_outputs, input_weights)
+            # (batch_size, enc_output_dim + enc_embedding_dim)
+            return torch.cat((attended_input, embedded_input), -1)
+        else:
+            return embedded_input
+
+    def _predict_step(self, # pylint: disable=no-self-use
+                      target_scores: torch.LongTensor) -> Dict[str, torch.LongTensor]:
+        """
+        Take the projected scores for all output classes and predict the output for a single timestep.
+        Here we simply normalize the scores and output the class with the highest probability as the
+        prediction. This method can be overridden to do a constrained prediction and/or avoid local
+        normalization.
+
+        Parameters
+        ----------
+        target_scores : torch.LongTensor
+            Unnormalized class scores, the output from a projection layer. (batch_size, num_classes)
+
+        Returns
+        -------
+        prediction_dict : Dict[str, torch.LongTensor] with fields
+            class_probabilities
+            predicted_classes
+        """
+        # (batch_size, num_classes)
+        class_probabilities = F.softmax(target_scores)
+        # (batch_size,)
+        _, predicted_classes = torch.max(class_probabilities, 1)
+        prediction_dict = {"class_probabilities": class_probabilities, "predicted_classes": predicted_classes}
+        return prediction_dict
+
+    @overrides
+    def decode(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
+        """
+        Trims the output predictions to the first end symbol, replaces indics with corresponding tokens,
+        and adds a field called ``predicted_tokens`` to the ``output_dict``.
+        """
+        predicted_indices = output_dict["predictions"]
+        if not isinstance(predicted_indices, numpy.ndarray):
+            predicted_indices = predicted_indices.cpu().numpy()
+        all_predicted_tokens = []
+        for indices in predicted_indices:
+            indices = list(indices)
+            # Collect indices till the first end_symbol
+            indices = indices[:indices.index(self._end_index)]
+            predicted_tokens = [self.vocab.get_token_from_index(x, namespace="output_tokens")
```

Comment:
This string doesn't match what's above.  Looks like you need a test for this.  And it should just be `self._target_namespace`, anyway.